backend/DynamicDispatch.cpp

```diff
@@ -126,22 +146,49 @@ Value *CodeGenerator::dynamicInvoke(const Node &node,
     Builder->SetInsertPoint(uniqueIdOkBB);
     vector<Value *> argVals;
     for(auto v : args) argVals.push_back(v.second);
+    /* The last arg is always function object, so that we can extract the closed overs inside the function */
+    argVals.push_back(objectToInvoke);
+
     Value *uniqueIdOkVal = Builder->CreateCall(staticFunctionToCall, argVals, string("call_dynamic"));
-    Builder->CreateBr(uniqueIdMergeBB);
     
+    Value *oncePtr = Builder->CreateStructGEP(runtimeFunctionType(), objectToInvoke, 3, "get_once");
+    Value *once = Builder->CreateLoad(Type::getInt8Ty(*TheContext), oncePtr, "load_once");
+    Value *onceCond = Builder->CreateIntCast(once, dynamicUnboxedType(booleanType), false);
+    
+    BasicBlock *cleanupBB = llvm::BasicBlock::Create(*TheContext, "cleanup", parentFunction);
+    BasicBlock *ccBB = llvm::BasicBlock::Create(*TheContext, "ccBB", parentFunction);
```

Comment:
Fragment generowanego kodu dla `fib.clj`:

<details>

```
type_function:                                    ; preds = %entry
  %get_unique_id = getelementptr inbounds %Function, ptr %load_var1, i32 0, i32 0
  %load_unique_id = load i64, ptr %get_unique_id, align 8
  %cmp_unique_id = icmp eq i64 %load_unique_id, 1
  br i1 %cmp_unique_id, label %unique_id_ok, label %unique_id_failed

unique_id_ok:                                     ; preds = %type_function
  %call_dynamic = call i64 @fn_1_J_J(i64 42, ptr %load_var1)
  %get_once = getelementptr inbounds %Function, ptr %load_var1, i32 0, i32 3
  %load_once = load i8, ptr %get_once, align 1
  %3 = trunc i8 %load_once to i1
  br i1 %3, label %cleanup, label %ccBB

cleanup:                                          ; preds = %unique_id_ok
  call void @Function_cleanupOnce(ptr %load_var1)
  br label %ccBB

ccBB:                                             ; preds = %cleanup, %unique_id_ok
  br label %unique_id_merge

unique_id_failed:                                 ; preds = %type_function
  %call_specialiseDynamicFn = call ptr @specialiseDynamicFn(ptr inttoptr (i64 140267768303616 to ptr), ptr %load_var1, i64 1, i64 1, i64 1, i64 0, i64 0, i64 0)
  %call_dynamic3 = call i64 %call_specialiseDynamicFn(i64 42, ptr %load_var1)
  %get_once4 = getelementptr inbounds %Function, ptr %load_var1, i32 0, i32 3
  %load_once5 = load i8, ptr %get_once4, align 1
  %4 = trunc i8 %load_once5 to i1
  br i1 %4, label %cleanup6, label %continue

cleanup6:                                         ; preds = %unique_id_failed
  call void @Function_cleanupOnce(ptr %load_var1)
  br label %continue

continue:                                         ; preds = %cleanup6, %unique_id_failed
  br label %unique_id_merge

unique_id_merge:                                  ; preds = %continue, %ccBB
  %phi = phi i64 [ %call_dynamic, %ccBB ], [ %call_dynamic3, %continue ]
  br label %merge
```

</details>

Tam są puste bloki, które nie mają żadnej instrukcji poza skokiem do następnego bloku. Nie wiem, na ile pozbycie się tych bloków przyspieszy czas wykonania. Tutaj na przykład, zamiast blok `ccBB` jest zbędny i z `unique_id_ok` i `cleanup` można skoczyć bezpośrednio do `unique_id_merge`, tak samo `continue`

Reply:
W wyspecjalizowanym `@fn_1_J_J`

<details>

```
define i64 @fn_1_J_J(i64 %0, ptr %1) {
entry:
  %cmp_jj_tmp = icmp eq i64 %0, 1
  br i1 %cmp_jj_tmp, label %then, label %else

then:                                             ; preds = %entry
  br label %ifcont8

else:                                             ; preds = %entry
  %cmp_jj_tmp1 = icmp eq i64 %0, 2
  br i1 %cmp_jj_tmp1, label %then2, label %else3

then2:                                            ; preds = %else
  br label %ifcont

else3:                                            ; preds = %else
  %sub_ii_tmp = sub i64 %0, 1
  %load_var = load atomic ptr, ptr @cv_user_fib monotonic, align 8
  %2 = getelementptr i8, ptr %load_var, i64 -16
  %get_type = getelementptr inbounds %Object, ptr %2, i32 0, i32 1
  %3 = atomicrmw volatile add ptr %get_type, i64 1 monotonic, align 8
  %call_fn_1_J = call i64 @fn_1_J_J(i64 %sub_ii_tmp, ptr %load_var)
  %sub_ii_tmp4 = sub i64 %0, 2
  %load_var5 = load atomic ptr, ptr @cv_user_fib monotonic, align 8
  %4 = getelementptr i8, ptr %load_var5, i64 -16
  %get_type6 = getelementptr inbounds %Object, ptr %4, i32 0, i32 1
  %5 = atomicrmw volatile add ptr %get_type6, i64 1 monotonic, align 8
  %call_fn_1_J7 = call i64 @fn_1_J_J(i64 %sub_ii_tmp4, ptr %load_var5)
  %add_ii_tmp = add i64 %call_fn_1_J, %call_fn_1_J7
  br label %ifcont

ifcont:                                           ; preds = %then2, %else3
  %iftmp = phi i64 [ 1, %then2 ], [ %add_ii_tmp, %else3 ]
  br label %ifcont8

ifcont8:                                          ; preds = %then, %ifcont
  %iftmp9 = phi i64 [ 1, %then ], [ %iftmp, %ifcont ]
  %6 = getelementptr i8, ptr %1, i64 -16
  %get_type10 = getelementptr inbounds %Object, ptr %6, i32 0, i32 1
  %7 = atomicrmw volatile sub ptr %get_type10, i64 1 monotonic, align 8
  %cmp_jj = icmp eq i64 %7, 1
  br i1 %cmp_jj, label %destroy, label %ignore

destroy:                                          ; preds = %ifcont8
  call void @Object_destroy(ptr %6, i8 1)
  br label %release_cont

ignore:                                           ; preds = %ifcont8
  br label %release_cont

release_cont:                                     ; preds = %ignore, %destroy
  %release_phi = phi i1 [ true, %destroy ], [ false, %ignore ]
  ret i64 %iftmp9
}
```

</details>

Co prawda wołasz od razu wyspecjalizowaną funkcję, zamiast ładować vara od nowa, ale wciąż blok `ignore` jest bez sensu.

Reply:
Włączyłem aby wypisywał te bloki po optymalizacji i problem wydaje się że znika. Mi było trudno to napisać bez nich, ale myślę że skoro optymalizator to i tak wywala to może może tak być.